infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/UniqueWorkspacePVCStrategy.java

```diff
@@ -106,67 +121,143 @@ public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
-        k8sClaims.waitBound(pvc.getMetadata().getName(), timeoutMillis);
+    if (k8sEnv.getPersistentVolumeClaims().isEmpty()) {
+      // no PVCs to prepare
+      return;
+    }
+
+    final KubernetesPersistentVolumeClaims k8sClaims =
+        factory.create(workspaceId).persistentVolumeClaims();
+    LOG.debug("Creating PVCs for workspace '{}'", workspaceId);
+    k8sClaims.createIfNotExist(k8sEnv.getPersistentVolumeClaims().values());
+
+    LOG.debug("Waiting PVCs for workspace '{}' to be bound", workspaceId);
+    for (PersistentVolumeClaim pvc : k8sEnv.getPersistentVolumeClaims().values()) {
+      k8sClaims.waitBound(pvc.getMetadata().getName(), timeoutMillis);
+    }
+    LOG.debug("Preparing PVCs done for workspace '{}'", workspaceId);
+  }
+
+  @Override
+  public void cleanup(Workspace workspace) throws InfrastructureException {
+    if (EphemeralWorkspaceUtility.isEphemeral(workspace)) {
+      return;
+    }
+    String workspaceId = workspace.getId();
+    factory
+        .create(workspaceId)
+        .persistentVolumeClaims()
+        .delete(ImmutableMap.of(CHE_WORKSPACE_ID_LABEL, workspaceId));
+  }
+
+  private void processUserDefinedPVCs(
+      KubernetesEnvironment k8sEnv,
+      RuntimeIdentity identity,
+      String workspaceId,
+      Map<String, PersistentVolumeClaim> volumeName2PVC) {
+    // process user-defined PVCs according to unique strategy
+    final Map<String, PersistentVolumeClaim> envClaims = k8sEnv.getPersistentVolumeClaims();
+    Map<String, PersistentVolumeClaim> userDefinedPVCs =
+        envClaims
+            .values()
+            .stream()
+            .filter(
+                p -> {
+                  Object isProvisioned = p.getAdditionalProperties().get("CHE_PROVISIONED");
+                  return !(isProvisioned instanceof Boolean) || !(Boolean) isProvisioned;
+                })
+            .collect(toMap(pvc -> pvc.getMetadata().getName(), Function.identity()));
+
+    prefixSubpaths(userDefinedPVCs.keySet(), k8sEnv.getPodsData(), identity.getWorkspaceId());
+
+    for (PersistentVolumeClaim pvc : userDefinedPVCs.values()) {
+      String originalPVCName = pvc.getMetadata().getName();
+
+      PersistentVolumeClaim existingPVC = volumeName2PVC.get(originalPVCName);
+
+      if (existingPVC != null) {
+        // Replace pvc in environment with existing. Fix the references in Pods
+        envClaims.remove(originalPVCName);
+        changePVCReferences(
+            k8sEnv.getPodsData(), originalPVCName, existingPVC.getMetadata().getName());
+      } else {
+        // there is no the corresponding existing pvc
+        // new one should be created with generated name
+        putLabel(pvc, CHE_VOLUME_NAME_LABEL, originalPVCName);
+        putLabel(pvc, CHE_WORKSPACE_ID_LABEL, workspaceId);
+
+        final String uniqueName = Names.generateName(pvcNamePrefix + '-');
+        pvc.getMetadata().setName(uniqueName);
+        pvc.getAdditionalProperties().put("CHE_PROVISIONED", true);
```

Comment:
We should pull `"CHE_PROVISIONED"` out into a constant to avoid typos.

Reply:
Done